<div id='top'>包括以下操作：</div>
<li><a href='#1'>CrossEntropy</a></li>
<li><a href='#2'>InfoNCE</a></li>
<li><a href='#3'></a></li>
<li><a href='#4'></a></li>
<li><a href='#5'></a></li>
<li><a href='#6'></a></li>

<div style='color:skyblue; font-size:24px' id='1'>CrossEntropy</div>

<a href='#top'>▲ Top</a>

在交叉熵损失的数学定义中，真实标签确实可以理解为隐式的独热编码（one-hot encoding），但在实际代码实现时（如PyTorch），无需显式将标签转换为独热编码，直接用类别索引即可。

![entropy.png](entropy.png)

In [ ]:
import torch
from math import log
logits = torch.tensor([[-0.5651, 0.2735, 0.2883, 0.7495, 0.4461],
                       [-0.4567, 0.9265, 0.8855, 0.5502, 1.2194]])
label = torch.tensor([1, 2]) # 标签 [1, 2] 是类别索引，无需手动转换为独热编码。PyTorch内部会高效处理

In [11]:
# loss = torch.nn.CrossEntropyLoss(reduction='none')
loss = torch.nn.CrossEntropyLoss()
output = loss(logits, label)
print(output)

tensor(1.5711)


In [ ]:
m = torch.nn.LogSoftmax(dim=-1)
b = m(logits) # 得到归一化后的值
print(b)
nl = torch.nn.NLLLoss() # 实例化对象时，可以选择使用weight,size_average 等参数
n_loss = nl(b,label) # 这里面是要和 NLLLoss 类中forward() 方法中的参数保持一致，也就是需要传入 input 和 target 
print(n_loss)

tensor([[-2.4959, -1.6573, -1.6425, -1.1813, -1.4847],
        [-2.8271, -1.4439, -1.4849, -1.8202, -1.1510]])
tensor(1.5711)


In [ ]:
m = torch.nn.Softmax(dim=-1)
Softmax_b = m(logits) # 得到归一化后的值
print(Softmax_b)
Softmax_log = torch.log(Softmax_b) # 取对数
print(Softmax_log)

nl = torch.nn.NLLLoss() # 实例化对象时，可以选择使用weight,size_average 等参数
n_loss = nl(Softmax_log,label) # 这里面是要和 NLLLoss 类中forward() 方法中的参数保持一致，也就是需要传入 input 和 target 
print(n_loss)

tensor([[0.0824, 0.1906, 0.1935, 0.3069, 0.2266],
        [0.0592, 0.2360, 0.2265, 0.1620, 0.3163]])
tensor([[-2.4959, -1.6573, -1.6425, -1.1813, -1.4847],
        [-2.8271, -1.4439, -1.4849, -1.8202, -1.1510]])
tensor(1.5711)


样本1：Label是1 → 对应概率 0.1906<br>
样本2：Label是2 → 对应概率 0.2265<br>
-log(0.1906) = 1.6573 <br>
-log(0.2265) = 1.4849<br>
loss = (1.6573 + 1.4849)/2 = 1.5711

<div style='color:skyblue; font-size:24px' id='2'>InfoNCE</div>

<a href='#top'>▲ Top</a>